### ISSUE TITLE
[xcode11-feature] Swift Libraries (CryptoKit and RealityKit)

### ISSUE THREAD

**Issue by chamons** [xamarin/xamarin-macios#6239]
Both [CryptoKit](https://developer.apple.com/documentation/cryptokit) and [RealityKit](https://developer.apple.com/documentation/realitykit) are Swift-only, with no Objective-C binding.

Infrastructure to work on swift is in work but not public currently.

If we wish to support these libraries, we need to figure out what shipping a Swift library binding truly entails. 

> As a friendly reminder, an issue card is not a guarantee that a certain feature will be done, or done under any particular time frame. Issues allow us to document what features we do and do not have, and give the community a place to provide feedback.

**Comment by wjk** [comment]
Just a comment, if binding to Swift natively is too difficult (although it should be easier now that the ABI has been fixed), we could always write a Swift library that exposes `@objc` entrypoints, and then call those from C# instead. Inelegant, but workable.

**Comment by softlion** [comment]
Any news on this ? Now that SceneKit is officially out, RealityKit is the way to go. And it's in swift only.

**Comment by chamons** [comment]
Hey folks, while not a full solution https://github.com/xamarin/xamarin-macios/issues/15315 might help resolve part of the issue, depending on what APIs you need.

**Comment by CastleSoft** [comment]
Any progress on RealityKit ?

https://developer.apple.com/augmented-reality/realitykit/ 

C# wrappers ?

Thanks

**Comment by chamons** [comment]
There is no current progress on this issue. However, using [this prototype](https://github.com/xamarin/xamarin-macios/issues/15315) you could potentially bind the subset of the API you need yourself.

**Comment by tipa** [comment]
This is now the 3rd library (after StoreKit2 & WidgetKit) that I wanted to use, just to find out it is not supported. Really hope there will be some progress is making Swift-APIs available...

**Comment by rolfbjarne** [comment]
> I just found out that in .NET8, the `AesGcm` class is using CryptoKit under the hood ([dotnet/runtime#29811](https://github.com/dotnet/runtime/issues/29811)), maybe the work that has been done there can be used to make CryptoKit available for direct use?

Not really, in dotnet/runtime this was partially implemented in Swift, by creating C-style functions in Swift that were then consumed from C# using DllImports.

In other words this is all custom, manually written, code. Exposing entire frameworks like this is quite labor intensive, and not feasible with the manpower we have available.

**Comment by mlancione** [comment]
Do we know whether Microsoft intends to ever support Swift Libraries? This seems quite critical to supporting the iOS platform going forward. 

Xamarin as a company seemed to make it their core mission to supporting all iOS and Android functionality (with same-day OS support as well).

**Comment by dalexsoto** [comment]
Yes this is on our radar and we intend to look into this in the future, I do not have a timeline for you just yet but it is something we will be actively looking into.

**Comment by mlancione** [comment]
@dalexsoto Thanks for the update. Much appreciated.

**Comment by karant0007** [comment]
> Yes this is on our radar and we intend to look into this in the future, I do not have a timeline for you just yet but it is something we will be actively looking into.

@dalexsoto @rolfbjarne 
Is there any update on this? i wanted to use it in .net maui, and it's necessary to use realityKit. 
I have tried slim binding but problem with slim binding is i have write code for each property i want to use which doesn't seems good. as i'm not only using methods but structs as well.

**Comment by rolfbjarne** [comment]
> > Yes this is on our radar and we intend to look into this in the future, I do not have a timeline for you just yet but it is something we will be actively looking into.
> 
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/204671?s=64&amp;u=6d5eeb050c1891baab90a953f7b4319f855b9c68&amp;v=4">@dalexsoto](https://github.com/dalexsoto) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/249268?s=64&amp;v=4">@rolfbjarne](https://github.com/rolfbjarne) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/jamesmontemagno?size=32">@jamesmontemagno](https://github.com/jamesmontemagno) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/jfversluis?size=32">@jfversluis](https://github.com/jfversluis) Is there any update on this? i wanted to use it in .net maui, and it's necessary to use realityKit. I have tried slim binding but problem with slim binding is i have write code for each property i want to use which doesn't seems good. as i'm not only using methods but structs as well.

Swift support is actively being worked on in dotnet/runtime (see https://github.com/dotnet/runtime/issues/95633 for some of it), but it's a _lot_ of work, so it'll take a while until we can actually use the Swift support for bindings (current estimate: not before .NET 11).

**Comment by karant0007** [comment]
> > > Yes this is on our radar and we intend to look into this in the future, I do not have a timeline for you just yet but it is something we will be actively looking into.
> > 
> > 
> > [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/204671?s=64&amp;u=6d5eeb050c1891baab90a953f7b4319f855b9c68&amp;v=4">@dalexsoto](https://github.com/dalexsoto) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/249268?s=64&amp;v=4">@rolfbjarne](https://github.com/rolfbjarne) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/jamesmontemagno?size=32">@jamesmontemagno](https://github.com/jamesmontemagno) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/jfversluis?size=32">@jfversluis](https://github.com/jfversluis) Is there any update on this? i wanted to use it in .net maui, and it's necessary to use realityKit. I have tried slim binding but problem with slim binding is i have write code for each property i want to use which doesn't seems good. as i'm not only using methods but structs as well.
> 
> Swift support is actively being worked on in dotnet/runtime (see [dotnet/runtime#95633](https://github.com/dotnet/runtime/issues/95633) for some of it), but it's a _lot_ of work, so it'll take a while until we can actually use the Swift support for bindings (current estimate: not before .NET 11).

@rolfbjarne 

Thanks for reply, 
I would like to contribute on this, as i required this asap in one my project. 
other wise we have to port our app to native or react native / Flutter. 

If you can share some details like how it works may be i can try to help

**Comment by luvkushparoksya-blip** [comment]
> > > > > Yes this is on our radar and we intend to look into this in the future, I do not have a timeline for you just yet but it is something we will be actively looking into.
> > > > 
> > > > 
> > > > [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/204671?s=64&amp;u=6d5eeb050c1891baab90a953f7b4319f855b9c68&amp;v=4">@dalexsoto](https://github.com/dalexsoto) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/249268?s=64&amp;v=4">@rolfbjarne](https://github.com/rolfbjarne) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/jamesmontemagno?size=32">@jamesmontemagno](https://github.com/jamesmontemagno) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/jfversluis?size=32">@jfversluis](https://github.com/jfversluis) Is there any update on this? i wanted to use it in .net maui, and it's necessary to use realityKit. I have tried slim binding but problem with slim binding is i have write code for each property i want to use which doesn't seems good. as i'm not only using methods but structs as well.
> > > 
> > > 
> > > Swift support is actively being worked on in dotnet/runtime (see [dotnet/runtime#95633](https://github.com/dotnet/runtime/issues/95633) for some of it), but it's a _lot_ of work, so it'll take a while until we can actually use the Swift support for bindings (current estimate: not before .NET 11).
> > 
> > 
> > [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/249268?s=64&amp;v=4">@rolfbjarne](https://github.com/rolfbjarne)
> > Thanks for reply, I would like to contribute on this, as i required this asap in one my project. other wise we have to port our app to native or react native / Flutter.
> > If you can share some details like how it works may be i can try to help
> 
> If you want to help, the best way would be to ask the runtime about it on that issue ([dotnet/runtime#95633](https://github.com/dotnet/runtime/issues/95633)).

Is there any update on this RealityKit ? @rolfbjarne @dalexsoto @jfversluis 
i want to use into .net maui, but its swift library and internally uses RealityKit

https://github.com/SceneView/sceneview

**Comment by softlion** [comment]
And issue https://github.com/dotnet/runtime/issues/95633 is marked as completed since 1 year.

**Comment by rolfbjarne** [comment]
> > > > > > Yes this is on our radar and we intend to look into this in the future, I do not have a timeline for you just yet but it is something we will be actively looking into.
> > > > > 
> > > > > 
> > > > > [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/204671?s=64&amp;u=6d5eeb050c1891baab90a953f7b4319f855b9c68&amp;v=4">@dalexsoto](https://github.com/dalexsoto) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/249268?s=64&amp;v=4">@rolfbjarne](https://github.com/rolfbjarne) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/jamesmontemagno?size=32">@jamesmontemagno](https://github.com/jamesmontemagno) [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/jfversluis?size=32">@jfversluis](https://github.com/jfversluis) Is there any update on this? i wanted to use it in .net maui, and it's necessary to use realityKit. I have tried slim binding but problem with slim binding is i have write code for each property i want to use which doesn't seems good. as i'm not only using methods but structs as well.
> > > > 
> > > > 
> > > > Swift support is actively being worked on in dotnet/runtime (see [dotnet/runtime#95633](https://github.com/dotnet/runtime/issues/95633) for some of it), but it's a _lot_ of work, so it'll take a while until we can actually use the Swift support for bindings (current estimate: not before .NET 11).
> > > 
> > > 
> > > [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/249268?s=64&amp;v=4">@rolfbjarne](https://github.com/rolfbjarne)
> > > Thanks for reply, I would like to contribute on this, as i required this asap in one my project. other wise we have to port our app to native or react native / Flutter.
> > > If you can share some details like how it works may be i can try to help
> > 
> > 
> > If you want to help, the best way would be to ask the runtime about it on that issue ([dotnet/runtime#95633](https://github.com/dotnet/runtime/issues/95633)).
> 
> Is there any update on this RealityKit ? [@rolfbjarne](https://github.com/rolfbjarne) [@dalexsoto](https://github.com/dalexsoto) [@jfversluis](https://github.com/jfversluis) i want to use into .net maui, but its swift library and internally uses RealityKit
> 
> [SceneView/sceneview](https://github.com/SceneView/sceneview)

There hasn't been any progress on our side yet, but someone from the community built something you might find useful: https://www.reddit.com/r/dotnetMAUI/comments/1s49uf3/introducing_swift_dotnet_bindings_swift_objc/
